action/protocol/staking/read_state.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2020 IoTeX Foundation
+// This is an alpha (internal) release and is not suitable for production. This source code is provided 'as is' and no
+// warranties are given as to title or non-infringement, merchantability or fitness for purpose and, to the extent
+// permitted by law, all liability for your use of the code is disclaimed. This source code is governed by Apache
+// License 2.0 that can be found in the LICENSE file.
+
+package staking
+
+import (
+	"context"
+
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/iotexproject/iotex-core/action/protocol"
+	"github.com/iotexproject/iotex-core/state"
+	"github.com/iotexproject/iotex-proto/golang/iotexapi"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"github.com/pkg/errors"
+)
+
+func (p *Protocol) readStateBuckets(ctx context.Context, sr protocol.StateReader,
+	req *iotexapi.ReadStakingDataRequest_VoteBuckets) (*iotextypes.VoteBucketList, error) {
+	all, err := getAllBuckets(sr)
+	if err != nil {
+		return nil, err
+	}
+
+	offset := int(req.GetPagination().GetOffset())
+	limit := int(req.GetPagination().GetLimit())
+	buckets := getPageOfBuckets(all, offset, limit)
+	return toIoTexTypesVoteBucketList(buckets)
+}
+
+func (p *Protocol) readStateBucketsByVoter(ctx context.Context, sr protocol.StateReader,
+	req *iotexapi.ReadStakingDataRequest_VoteBucketsByVoter) (*iotextypes.VoteBucketList, error) {
+	voter, err := address.FromString(req.GetVoterAddress())
+	if err != nil {
+		return nil, err
+	}
+	indices, err := getVoterBucketIndices(sr, voter)
+	if errors.Cause(err) == state.ErrStateNotExist {
+		return &iotextypes.VoteBucketList{}, nil
+	}
+	if indices == nil || err != nil {
+		return nil, err
+	}
+	buckets, err := getBucketsWithIndices(sr, *indices)
+	if err != nil {
+		return nil, err
+	}
+
+	offset := int(req.GetPagination().GetOffset())
+	limit := int(req.GetPagination().GetLimit())
+	buckets = getPageOfBuckets(buckets, offset, limit)
+	return toIoTexTypesVoteBucketList(buckets)
+}
+
+func (p *Protocol) readStateBucketsByCandidate(ctx context.Context, sr protocol.StateReader,
+	req *iotexapi.ReadStakingDataRequest_VoteBucketsByCandidate) (*iotextypes.VoteBucketList, error) {
+	cands, err := getAllCandidates(sr)
```

Comment:
you can add a func
func getCandidateByAddr(sr protocol.StateReader, addr address.Address) (Candidate, error)
and avoid filtering below.

i will implement that soon

Reply:
you will need to first map the name --> address from CandidateCenter

Reply:
let's optimize in next PR

Reply:
done